### ISSUE TITLE
wgpu validation error when using 16bits precision PNGs

### ISSUE THREAD

**Issue by linkpy** [bevyengine/bevy#4532]
## Bevy version

0.7.0

## Operating system & version

Pop_OS! 21.10

## What you did

I tried to use a normal texture which has a precision of "16-bit gamma integer".

## What you expected to happen

I've expected to have the texture loaded normally and used with a `PbrBundle`.

## What actually happened

Crash at runtime.

```
2022-04-19T09:57:07.813153Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1    
2022-04-19T09:57:07.885682Z  INFO bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce GTX 1660", vendor: 4318, device: 8580, device_type: DiscreteGpu, backend: Vulkan }
2022-04-19T09:57:08.200302Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default    
thread 'TaskPool (7)' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
      note: label = `pbr_standard_material_bind_group`
    texture binding 9 expects sample type = Float { filterable: true }, but given a view with format = Rgba16Uint

', /home/pancake/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2273:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'task has failed', /home/pancake/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.2.0/src/task.rs:425:45
```

## Additional information

I've compared my texture with other textures which work correctly : 

![image](https://user-images.githubusercontent.com/2700596/163980082-fefc002a-42b7-479a-9c42-c68b50a70e43.png)

Left is the image used by https://github.com/bevyengine/bevy/blob/main/examples/3d/texture.rs and right is the normal texture generating the wgpu validation error. The only difference I found was the precision.

After re-exporting the image to use 8-bit precision, the error is gone.

**Comment by superdump** [comment]
I wonder if this should maybe use `Rgba16Unorm`.

The PNG image loader only knows that the components are 16-bit unsigned integer, it doesn't know how you want to use it. The PBR material shader binding is configured to always expect a floating point value, whether that is how it is stored, or how it is converted automatically when being sampled.

So, you should be able to load the texture, and when it is loaded, reconfigure it:
```rust
if let Some(image) = images.get_mut(image_handle) {
    image.texture_descriptor.format = TextureFormat::Rgba16Unorm;
}
```
If this is part of a model, you may have to delay spawning the model into the world until this change is made.

In future, these friction points would be addressed by a few things:
- image assets being used for normal map textures would automatically have their texture format reconfigured to a float mapping format, where necessary
- asset meta files would allow configuration of what texture format to use when the underlying data is in the same format, but the format can be converted to something else such as unsigned normalised float when sampling

@cart I think this is an interesting case to consider for asset stuff so just giving you a nudge to inject it into your brain. :)

**Comment by superdump** [comment]
Changes were made to texture stuff between 0.6 and 0.7. I don't think anything should have been changed with respect to PNG textures but perhaps it was. I guess I/someone will have to go digging to find out what changed and why.

Some more information would likely be useful - exactly how are you loading the texture? Is it part of a glTF model or some other kind of model?

**Comment by Sheepyhead** [comment]
> Changes were made to texture stuff between 0.6 and 0.7. I don't think anything should have been changed with respect to PNG textures but perhaps it was. I guess I/someone will have to go digging to find out what changed and why.
> 
> Some more information would likely be useful - exactly how are you loading the texture? Is it part of a glTF model or some other kind of model?

For my case it's a texture that's part of a .gltf file I load as a scene. Some meshes are fully textured with UV maps etc., and exported from Blender. In 0.6.1 I didn't get any panics loading this scene (although it did look strange, something I suspect is an unrelated lighting issue)

**Comment by superdump** [comment]
> > Changes were made to texture stuff between 0.6 and 0.7. I don't think anything should have been changed with respect to PNG textures but perhaps it was. I guess I/someone will have to go digging to find out what changed and why.
> > Some more information would likely be useful - exactly how are you loading the texture? Is it part of a glTF model or some other kind of model?
> 
> For my case it's a texture that's part of a .gltf file I load as a scene. Some meshes are fully textured with UV maps etc., and exported from Blender. In 0.6.1 I didn't get any panics loading this scene (although it did look strange, something I suspect is an unrelated lighting issue)

Normals are essential for lighting and have a direct impact. It could be related. Could you share the complete model so I can debug it?

**Comment by Sheepyhead** [comment]
> Normals are essential for lighting and have a direct impact. It could be related. Could you share the complete model so I can debug it?

Yeah it's this model here https://github.com/Sheepyhead/diablo_3/blob/main/assets/GameScene.gltf

**Comment by Sheepyhead** [comment]
> i realized the repo is private, I'll open it when I can get to my computer

It should now be public

**Comment by bofh** [comment]
Hello,

I got a similar error with 16 bit RGBA PNGs on Bevy 0.7.0 in my code. So I use the [pngtest16rgba.png](https://www.fnordware.com/superpng/pngtest16rgba.png) in `examples/2d/sprite.rs` to exclude my code and PNG and got:

```
2022-06-30T11:24:00.227870Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1    
2022-06-30T11:24:00.437052Z  INFO bevy_render::renderer: AdapterInfo { name: "Quadro GP100", vendor: 4318, device: 5616, device_type: DiscreteGpu, backend: Vulkan }
2022-06-30T11:24:00.704090Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default    
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
      note: label = `sprite_material_bind_group`
    texture binding 0 expects sample type = Float { filterable: true }, but given a view with format = Rgba16Uint

', /home/dev/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2273:5
```

**Comment by alexniver** [comment]
same issue in 0.8.0


```
2022-08-22T11:51:24.068760Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
      note: label = `sprite_material_bind_group`
    texture binding 0 expects sample type = Float { filterable: true }, but given a view with format = Rgba16Uint

', C:\Users\zhuli\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-0.13.1\src\backend\direct.rs:2391:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\release\examples\sprite.exe` (exit code: 101)

```

PNG file exported by krita

![捕获](https://user-images.githubusercontent.com/3366229/185916481-2cbb6d3b-3a57-45fa-852c-5f358f1e45d9.PNG)

**Comment by bofh** [comment]
Related: #554 and #1824

**Comment by jancespivo** [comment]
In 0.11.2 there is the same issue as well. I tried to change color depth for all 16bit/color materials to 8bit/color via imagemagick (`magick <input image> -depth 8 <output image>`) and the model was then partially rendered (parts with affected textures were missing). Don't know why :shrug:

In main branch on commit bdb063497d802b9180d1b608bf54339e5a726b70 it seems the behavior is: meshes with textures with 16bit/color are not rendered (like in my attempt to fix it above). However it doesn't panic.

**Comment by pcwalton** [comment]
I *think* 6c1f4668c7afe20a6ba0657794a4cf51c803c0b8 fixed this.

**Comment by mirsella** [comment]
still crashing on wasm as of bevy 0.15.1
